plugin/pkg/policy/rqdata.go

```diff
@@ -0,0 +1,215 @@
+package policy
+
+import (
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+type requestFunc func(state request.Request) string
+
+// RequestExtractorMapping define the mapping between 'name' of data and the way to extract that data from the Request
+// it also defines what will be the empty value returned if the data behind the name is empty.
+// it is pretty static, and you should need to instantiate only once
+type RequestExtractorMapping struct {
+	replacements map[string]requestFunc
+	emptyValue   string
+}
+
+// RequestDataExtractor implements a GetValue(name) (value, valid) function
+// which allow to extract data from an existing DNS Request(or state)
+type RequestDataExtractor struct {
+	state     request.Request
+	requester *RequestExtractorMapping
+}
+
+// NewRequestExtractorMapping build the mapping name -> func to extract data from the Request
+func NewRequestExtractorMapping(emptyValue string) *RequestExtractorMapping {
```

Comment:
this looks super expensive, is this done for all request? Why is this needed?

Reply:
It is expensive. But it is created only once at setup of Firewall.
Then, on each query, the map is associated with the State of the query for extraction of the data - if needed by the expression